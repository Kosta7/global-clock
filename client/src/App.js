import React, { Component } from "react";
import styled from "styled-components";

import { ClockList } from "./ClockList";
import { ControlPanel } from "./ControlPanel";

const BACKEND_PATH = "/clocks";

class App extends Component {
  state = {
    editMode: false,
    clockList: []
  };

  componentDidMount() {
    this.updateClockList();
  }

  render() {
    return (
      <StyledApp>
        <ClockList
          clockList={this.state.clockList}
          addClock={this.addClock.bind(this)}
          deleteClock={this.deleteClock.bind(this)}
          editMode={this.state.editMode}
          onRef={ref => (this.clockList = ref)}
        />
        <ControlPanel
          editMode={this.state.editMode}
          toggleEditMode={this.toggleEditMode.bind(this)}
          resetShift={this.resetShift.bind(this)}
        />
      </StyledApp>
    );
  }

  resetShift() {
    this.clockList.scrollToNow();
  }

  toggleEditMode() {
    this.setState({ editMode: !this.state.editMode });
    this.resetShift();
  }

  async updateClockList() {
    const clockList = await this.fetchClockList();
    this.setState({ clockList });
  }

  async fetchClockList() {
    try {
      const data = await fetch(BACKEND_PATH);
      return await data.json();
    } catch (err) {
      console.log(err);
      return [];
    }
  }

  async addClock(city, timezone) {
    await fetch(`${BACKEND_PATH}?city=${city}&timezone=${timezone}`, {
      method: "POST"
    });

    // or return something from fetch? or just setstate?
    this.updateClockList();
  }

  async deleteClock(id) {
    await fetch(`${BACKEND_PATH}/${id}`, { method: "DELETE" });

    const clockList = this.state.clockList.filter(clock => clock.id !== id);
    this.setState({ clockList });
  }
}

const StyledApp = styled.div`
  position: relative;
  background-color: #cddcf0;
  font-family: monospace;
`;

export default App;
