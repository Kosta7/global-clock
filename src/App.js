import React, { Component } from "react";
import styled from "styled-components";

import { ClockList } from "./ClockList";
import { ControlPanel } from "./ControlPanel";

class App extends Component {
  state = {
    shift: 0,
    editMode: false
  };

  render() {
    const { shift, editMode } = this.state;

    return (
      <StyledApp>
        <ClockList shift={shift} editMode={editMode} />
        <ControlPanel
          isChangeShiftDisabled={editMode}
          isResetShiftDisabled={shift === 0}
          editMode={editMode}
          incrementShift={this.incrementShift.bind(this)}
          decrementShift={this.decrementShift.bind(this)}
          resetShift={this.resetShift.bind(this)}
          toggleEditMode={this.toggleEditMode.bind(this)}
        />
      </StyledApp>
    );
  }

  incrementShift() {
    this.setState({ shift: this.state.shift + 1 });
  }

  decrementShift() {
    this.setState({ shift: this.state.shift - 1 });
  }

  resetShift() {
    this.setState({ shift: 0 });
  }

  toggleEditMode() {
    this.setState({ editMode: !this.state.editMode });
    this.resetShift();
  }
}

const StyledApp = styled.div`
  position: relative;
  background-color: #cddcf0;
  min-height: 100vh;
`;

export default App;
