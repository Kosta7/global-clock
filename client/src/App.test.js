import React from "react";
import { mount, shallow } from "enzyme";

import App from "./App.js";

it("renders correctly", () => {
  const wrapper = shallow(<App />);

  expect(wrapper).toMatchSnapshot();
});

it("toggles edit mode", () => {
  const wrapper = mount(<App />);
  const instance = wrapper.instance();
  expect(instance.state.isEditMode).toEqual(false);
  instance.toggleEditMode();
  expect(instance.state.isEditMode).toEqual(true);
});

it("toggles edit mode", () => {
  const wrapper = mount(<App />);
  const instance = wrapper.instance();
  expect(instance.state.isEditMode).toEqual(false);
  instance.toggleEditMode();
  expect(instance.state.isEditMode).toEqual(true);
});
