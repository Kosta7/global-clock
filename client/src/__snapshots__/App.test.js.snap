// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<styled.div>
  <Styled(Header) />
  <ClockList
    addClock={[Function]}
    clockList={Array []}
    deleteClock={[Function]}
    isEditMode={false}
    isLoggedIn={false}
    onRef={[Function]}
  />
  <ControlPanel
    isEditMode={false}
    isLoggedIn={false}
    resetShift={[Function]}
    toggleEditMode={[Function]}
    updateClockList={[Function]}
  />
  <Loader />
</styled.div>
`;
